src/providers/azure.rs

```diff
@@ -0,0 +1,503 @@
+        self
+    }
+}
+
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). 
+///
+/// It also assumes you have previously run `az login` to authenticate with Microsoft.
+#[derive(Default)]
+pub struct AzureLauncher {
+    regions: HashMap<Region, AzureRegion>,
+}
+
+impl super::Launcher for AzureLauncher {
+    type MachineDescriptor = Setup;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        let region = l.region;
+        let mut az_region = AzureRegion::new(&l.region.to_string(), l.log.clone())?;
+        az_region.launch(l)?;
+        self.regions.insert(region, az_region);
+        Ok(())
+    }
+
+    fn connect_all<'l>(&'l self) -> Result<HashMap<String, crate::Machine<'l>>, Error> {
+        collect!(self.regions)
+    }
+}
+
+impl std::ops::Drop for AzureLauncher {
+    fn drop(&mut self) { }
+}
+
+struct Descriptor {
+    name: String,
+    ip: String,
+}
+
+/// Region-specific connection to Azure. Each instance of this type creates one Azure
+/// "resource group" and deletes the group on drop.
+///
+/// This implementation relies on the [Azure
+/// CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). 
+///
+/// It also assumes you have previously run `az login` to authenticate with Microsoft.
+#[derive(Default)]
+pub struct AzureRegion {
+    pub log: Option<slog::Logger>,
+    pub region: Region,
+    resource_group_name: String,
+    machines: Vec<Descriptor>,
+}
+
+impl AzureRegion {
+    pub fn new(region: &str, log: slog::Logger) -> Result<Self, Error> {
+        let region = region.parse()?;
+        let rg_name = super::rand_name("resourcegroup");
+
+        azcmd::create_resource_group(region, &rg_name)?;
+
+        Ok(Self {
+            log: Some(log),
+            region,
+            resource_group_name: rg_name,
+            machines: vec![],
+        })
+    }
+}
+
+impl super::Launcher for AzureRegion {
+    type MachineDescriptor = Setup;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        self.log = Some(l.log);
+        let log = self.log.as_ref().unwrap();
+        let max_wait = l.max_wait;
+        self.machines = 
+        l.machines
+            .into_iter()
+            .map(|(nickname, desc)| {
+                let vm_name = super::rand_name_sep("vm", "-");
+                debug!(log, "setting up azure instance"; "nickname" => &nickname, "vm_name" => &vm_name);
+
+                let pub_ip = azcmd::create_vm(
+                    &self.resource_group_name,
+                    &vm_name,
+                    &desc.instance_type,
+                    "ubuntu",
+                )?;
+
+                azcmd::open_ports(&self.resource_group_name, &vm_name)?;
+
+                if let Setup { setup_fn: Some(f), .. } = desc {
```

Comment:
This means setup will be run on one machine at a time, which will make setup of many machines very slow.